### PR TITLE
Less Lazy Assets

### PR DESCRIPTION
**Pull request by wyattjoh**
## What does this PR do?

Improves the asset fetching algorithm to attempt a fetch before the upsert. This will result in less writes to the database as a result.

**Note to 3.8.1 users, this rewrites some of the indexes found in that version with sparse indexes** (to improve performance) **and must be removed prior to activating this on production.**

### CODE REVIEW

File: graph/loaders/comments.js [hunk 1]
```diff
@@ -82,54 +82,52 @@ const getParentCountsByAssetID = (context, asset_ids) => {
...
+    (ctx.user == null || !ctx.user.can(SEARCH_NON_NULL_OR_ACCEPTED_COMMENTS))
   ) {
     return null;
   }
 
-  if (statuses && statuses.length > 0) {
-    query = query.where({status: {$in: statuses}});
-  }
-
-  if (ids) {
```
Comment (alinhle) [inline graph/loaders/comments.js:107]: This is not required anymore?

Comment (wyattjoh) [reply graph/loaders/comments.js:107]: We don't actually use that query anymore no. That was used before when we had to search for all the flags and then find all the comments from those flags with the given id's. The query paths in place now never use the `ids`.

File: graph/loaders/assets.js [hunk 2]
```diff
@@ -54,58 +37,121 @@ const getAssetsByQuery = async (context, query) => {
...
+  // Try to validate that the url is valid. If the URL constructor throws an
+  // error, throw our internal ErrInvalidAssetURL instead. This will validate
+  // that the url contains a valid scheme.
+  try {
+    new URL(url);
+  } catch (err) {
+    throw ErrInvalidAssetURL;
   }
 
-  let asset = await AssetsService.findOrCreateByUrl(asset_url);
```
Comment (alinhle) [inline graph/loaders/assets.js:71]: Whats the reason behind not implementing the changes in `AssetsService.findOrCreateByUrl`. Is it still required?

File: graph/loaders/comments.js [hunk 3]
```diff
@@ -82,54 +82,52 @@ const getParentCountsByAssetID = (context, asset_ids) => {
...
-    query = query.where({id: {$in: ids}});
-  }
+  const query = CommentModel.find();
 
   if (asset_id != null) {
-    query = query.where({asset_id});
+    query.merge({asset_id});
```
Comment (alinhle) [inline graph/loaders/comments.js:106]: Is this `merge` function documented somewhere? couldnt find it on the www.

**Comment by alinhle** [comment]
What happens if the old indexes are not removed? It won't run?

**Review by alinhle — APPROVED** [review]
LGTM

**Pull request merged**